nextcord/http.py

```diff
@@ -122,25 +138,331 @@ def bucket(self) -> str:
+        self.limit = 1 if x_limit is None else int(x_limit)
+
+        # Updates the remaining left if it exists, being pessimistic.
+        x_remaining = response.headers.get("X-RateLimit-Remaining")
+
+        if x_remaining is None:
+            self.remaining = 1
+        elif self._first_update:
+            self.remaining = int(x_remaining)
+        else:
+            # If requests come back out of order, it's possible that we could get a wrong amount remaining.
+            # It's best to be pessimistic and assume it cannot go back up unless the reset task occurs.
+            self.remaining = (
+                int(x_remaining) if int(x_remaining) < self.remaining else self.remaining
+            )
+
+        # Updates the datetime of the reset.
+        x_reset = response.headers.get("X-RateLimit-Reset")
+        if x_reset is not None:
+            x_reset = datetime.fromtimestamp(float(x_reset) + self._time_offset, tz=timezone.utc)
+
+            if self._use_reset_timestamp and self.reset is not None:
+                new_reset_diff = (x_reset - self.reset).total_seconds()
+                if self._tracked_reset_time < new_reset_diff:
+                    # Any time the tracked reset time is increased, the reset timestamp should also be increased
+                    #  thus we shouldn't have to (re)start the reset task here.
+                    self._tracked_reset_time = new_reset_diff
+
+            if self.reset is None or self.reset < x_reset:
+                self.reset = x_reset
+                if self._use_reset_timestamp:
+                    _log.debug(
+                        "Bucket %s: Reset timestamp increased, starting/resetting reset task.",
+                        self.bucket,
+                    )
+                    self.start_reset_task()
+
+        # Updates the reset-after count, being pessimistic.
+        x_reset_after = response.headers.get("X-RateLimit-Reset-After")
+        if x_reset_after is not None:
+            x_reset_after = float(x_reset_after) + self._time_offset
+            # TODO: Due to Discord constantly adjusting reset_after and requests not always being responded to in
+            #  order, self.reset_after can be incorrect and not representative of the rate limit state. Is it worth
+            #  doing more comparisons and work to make it more correct when possible, or is this good enough?
+            self.reset_after = x_reset_after
+            # Once we figure out what the true reset delay is, for example 5 seconds, we want to keep it at that.
+            if not self._use_reset_timestamp and self._tracked_reset_time < x_reset_after:
+                self._tracked_reset_time = x_reset_after
+                _log.debug(
+                    "Bucket %s: Reset after time increased to %s seconds, adapting reset time.",
+                    self.bucket,
+                    self._tracked_reset_time,
+                )
+                self.start_reset_task()
```

Comment:
Given from my understanding this is adaptive to possible future, changes seems fine as is?

Reply:
Agreed. Removing TODO.